### PR TITLE
Struct mech/truss hyperelastic laws

### PR DESCRIPTION
**Pull request by KlausBSautter**
New hyperelastic laws for the truss/cable element added including tests

### CODE REVIEW

**Comment by KlausBSautter** [comment]
@philbucher

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.hpp [hunk 1]
```diff
@@ -195,6 +195,8 @@ namespace Kratos
 
         Vector CalculateProjectionLengths();
 
+        double ReturnTangentModulus1D(const ProcessInfo& rCurrentProcessInfo) const;
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.hpp:198]: shouldn't this be a fct of the CLaw?

File: applications/StructuralMechanicsApplication/custom_constitutive/hyper_elastic_isotropic_henky_1d.h [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+
+    void FinalizeMaterialResponsePK2(Parameters& rValues) override
+    {
+        // plasticity law needs this function, so it is called in the truss element
+    };
+
+    //empty because called in the element and this base class throws an error
+    //if this is not overriden
+    void FinalizeNonLinearIteration(const Properties& rMaterialProperties,
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_constitutive/hyper_elastic_isotropic_henky_1d.h]: isn't this one of the deprecated functions?

File: applications/StructuralMechanicsApplication/custom_constitutive/hyper_elastic_isotropic_ogden_1d.h [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+
+    void FinalizeMaterialResponsePK2(Parameters& rValues) override
+    {
+        // plasticity law needs this function, so it is called in the truss element
+    };
+
+    //empty because called in the element and this base class throws an error
+    //if this is not overriden
+    void FinalizeNonLinearIteration(const Properties& rMaterialProperties,
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_constitutive/hyper_elastic_isotropic_ogden_1d.h]: this is deprecated

**Review by philbucher — COMMENTED** [review]
some comments

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.hpp (hunk 1 above)
Comment (KlausBSautter) [reply applications/CableNetApplication/custom_elements/sliding_cable_element_3D.hpp:198]: it is, this one just calls the claw

**Comment by KlausBSautter** [comment]
@philbucher i got rid of the `FinalizeNonLinearIteration`

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 4]
```diff
@@ -318,11 +318,11 @@ Vector SlidingCableElement3D::GetInternalForces()
   const int segments_number = points_number-1;
   const int points_int_number = points_number-2;
 
-  const double strain_gl      = this->CalculateGreenLagrangeStrain();
+  //const double strain_gl      = this->CalculateGreenLagrangeStrain();
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 5]
```diff
@@ -318,11 +318,11 @@ Vector SlidingCableElement3D::GetInternalForces()
...
   const double current_length = this->GetCurrentLength();
   const double prestress     = this->GetPK2PrestressValue();
   const double area           = this->GetProperties()[CROSS_AREA];
-  const double youngs_mod     = this->GetProperties()[YOUNG_MODULUS];
+  //const double youngs_mod     = this->GetProperties()[YOUNG_MODULUS];
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 6]
```diff
@@ -447,9 +447,7 @@ Matrix SlidingCableElement3D::ElasticStiffnessMatrix(const ProcessInfo& rCurrent
   const int dimension = 3;
   const SizeType local_size = dimension*points_number;
 
-  double youngs_mod = 0.00;
-  ConstitutiveLaw::Parameters Values(GetGeometry(),GetProperties(),rCurrentProcessInfo);
-  mpConstitutiveLaw->CalculateValue(Values,TANGENT_MODULUS,youngs_mod);
+  double youngs_mod = ReturnTangentModulus1D(rCurrentProcessInfo);
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]: `const`?

**Review by philbucher — DISMISSED** [review]

**Comment by KlausBSautter** [comment]
@philbucher could you re-approve?

**Review by philbucher — APPROVED** [review]

**Pull request merged**